Add a blog for building/publishing to container registries (#4686)
pulumi/docs: content/blog/build-publish-containers-iac/index.md

    if err != nil {
        return docker.ImageRegistry{}, err
    }
    parts := strings.Split(string(decoded), ":")
    if len(parts) != 2 {
        return docker.ImageRegistry{}, errors.New("Invalid credentials")
    }
    return docker.ImageRegistry{
        Server:   creds.ProxyEndpoint,
        Username: parts[0],
        Password: parts[1],
    }, nil
}).(docker.ImageRegistryOutput)
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Get registry info (creds and endpoint).
var imageName = repo.RepositoryUrl;
var registryInfo = repo.RegistryId.Apply(async (id) =>
{
    var creds = await GetCredentials.InvokeAsync(new GetCredentialsArgs { RegistryId = id });
    var decodedData = Convert.FromBase64String(creds.AuthorizationToken);
    var decoded = ASCIIEncoding.ASCII.GetString(decodedData);

    var parts = decoded.Split(':');
    if (parts.Length != 2)
    {
        throw new Exception("Invalid credentials");
    }

    return new ImageRegistry
    {
        Server = creds.ProxyEndpoint,
        Username = parts[0],
        Password = parts[1],
    };
});
```

{{< /choosable >}}

> #### Alternatievly, Authenticate with ECR Credential Helper

Reviewer: ```suggestion
#### Alternatively, Authenticate with ECR Credential Helper
```

Author: Good catch! Fixed.

    if err != nil {
        return docker.ImageRegistry{}, err
    }
    parts := strings.Split(string(decoded), ":")
    if len(parts) != 2 {
        return docker.ImageRegistry{}, errors.New("Invalid credentials")
    }
    return docker.ImageRegistry{
        Server:   creds.ProxyEndpoint,
        Username: parts[0],
        Password: parts[1],
    }, nil
}).(docker.ImageRegistryOutput)
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Get registry info (creds and endpoint).
var imageName = repo.RepositoryUrl;
var registryInfo = repo.RegistryId.Apply(async (id) =>
{
    var creds = await GetCredentials.InvokeAsync(new GetCredentialsArgs { RegistryId = id });
    var decodedData = Convert.FromBase64String(creds.AuthorizationToken);
    var decoded = ASCIIEncoding.ASCII.GetString(decodedData);

    var parts = decoded.Split(':');
    if (parts.Length != 2)
    {
        throw new Exception("Invalid credentials");
    }

    return new ImageRegistry
    {
        Server = creds.ProxyEndpoint,
        Username = parts[0],
        Password = parts[1],
    };
});
```

{{< /choosable >}}

#### Alternatively, Authenticate with ECR Credential Helper

> An alternative approach ECR supports is to [use the Docker credential helper from Amazon](https://aws.amazon.com/blogs/compute/authenticating-amazon-ecr-repositories-for-docker-cli-with-credential-helper/), which integrates with local IAM settings, adds smart caching, and removes the need for Docker to login. If you elect to go this route, simply leave out the username/password parts of the `docker.ImageRegistry`:

Reviewer 2: ```suggestion
An alternative approach ECR supports is to [use the Docker credential helper from Amazon](https://aws.amazon.com/blogs/compute/authenticating-amazon-ecr-repositories-for-docker-cli-with-credential-helper/), which integrates with local IAM settings, adds smart caching, and removes the need for Docker to login. If you elect to go this route, leave out the username/password parts of the `docker.ImageRegistry`:
```

{{< chooser language "javascript,typescript,python,go,csharp" / >}}

{{< choosable language javascript >}}

```javascript
// Get registry info (creds and endpoint).
var imageName = repo.repositoryUrl;
var registryInfo = undefined; // use ECR credentials helper.
```

{{< /choosable >}}

{{< choosable language typescript >}}

```typescript
// Get registry info (creds and endpoint).
const imageName = repo.repositoryUrl;
const registryInfo = undefined; // use ECR credentials helper.
```

{{< /choosable >}}

{{< choosable language python >}}

```python
# Get registry info (creds and endpoint).
image_name = repo.repository_url
registry_info = None # use ECR credentials helper.
```

{{< /choosable >}}

{{< choosable language go >}}

```go
// Get registry info (creds and endpoint).
imageName := repo.RepositoryUrl
registryInfo := docker.ImageRegistryArgs{} // use ECR credentials helper.
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Get registry info (creds and endpoint).
var imageName = repo.RepositoryUrl;
// Omit registryInfo and use ECR credentials helper.
```

{{< /choosable >}}

#### Learn More About ECR

> ECR offers many additional options not shown here. This includes configuring advanced IAM permissions, enabling image vulnerability scanning, managing the lifetime of images so that older, unused images are deleted based on configured policies, encrypting images, and more. For details on these advanced capabilities and more, refer to [Pulumi's ECR user guide](https://www.pulumi.com/docs/guides/crosswalk/aws/ecr/), [Pulumi's ECR API reference](https://www.pulumi.com/docs/reference/pkg/aws/ecr/repository/), or [Amazon's own product documentation](https://aws.amazon.com/ecr/).

Reviewer 2: ```suggestion
ECR offers many additional options not shown here. This includes configuring advanced IAM permissions, enabling image vulnerability scanning, managing the lifetime of images so that older, unused images are deleted based on configured policies, encrypting images, and more. For details on these advanced capabilities and more, refer to [Pulumi's ECR user guide](https://www.pulumi.com/docs/guides/crosswalk/aws/ecr/), [Pulumi's ECR API reference](https://www.pulumi.com/docs/reference/pkg/aws/ecr/repository/), or [Amazon's product documentation](https://aws.amazon.com/ecr/).
```

{{< /choosable >}}

{{< choosable language go >}}

```go
package main

import (
    "github.com/pulumi/pulumi-azure/sdk/v2/go/azure/containerservice"
    "github.com/pulumi/pulumi-azure/sdk/v2/go/azure/core"
    "github.com/pulumi/pulumi-docker/sdk/v2/go/docker"
    "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        // [Placeholder 1: Create a private ECR registry.]

        // [Placeholder 2: Get registry info (creds and endpoint).]

        // [Placeholder 3: Build and publish the container image.]
        return nil
    })
}        
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Pulumi;
using Pulumi.Azure.Core;
using Pulumi.Azure.ContainerService;
using Pulumi.Docker;

class Program
{
    static Task<int> Main() => Deployment.RunAsync(async () => {
        // [Placeholder 1: Create a private ECR registry.]

        // [Placeholder 2: Get registry info (creds and endpoint).]

        // [Placeholder 3: Build and publish the container image.]
    }
}
```

{{< /choosable >}}

#### Provision a New ACR Registry

> Next, declare a new ACR registry resource. To create a new registry, we need an Azure resource group; here we will create a new one, but feel free to use an existing one by simply passing its name in the resource group's place:

Reviewer 2: ```suggestion
Next, declare a new ACR registry resource. To create a new registry, we need an Azure resource group; here we will create a new one, but feel free to use an existing one by passing its name in the resource group's place:
```

}
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// New imports:
using Pulumi.Azure.Authorization;
using AAD = Pulumi.AzureAD;
using Pulumi.Random;

// [As before...]

// Get registry info (creds and endpoint).
var imageName = Output.Format($"{registry.LoginServer}/myapp");
var sp = new AAD.ServicePrincipal("mysp", new AAD.ServicePrincipalArgs
{
    ApplicationId = new AAD.Application("myspapp").ApplicationId,
});
var spPassword = new AAD.ServicePrincipalPassword("mysp-pass", new AAD.ServicePrincipalPasswordArgs
{
    ServicePrincipalId = sp.Id,
    Value = new RandomPassword("mypass",
        new RandomPasswordArgs
        {
            Length = 32,
            },
        new CustomResourceOptions { AdditionalSecretOutputs = { "result" } }
    ).Result,
    EndDateRelative = "8760h",
});
var spAuth = new Assignment("myauth", new AssignmentArgs
{
    Scope = registry.Id,
    RoleDefinitionName = "acrpush",
    PrincipalId = sp.Id,
});
var registryInfo = new Docker.ImageRegistry
{
    Server = registry.LoginServer,
    Username = sp.ApplicationId,
    Password = spAuth.Id.Apply(_ => spPassword.Value),
};
```

{{< /choosable >}}

#### Learn More About ACR

> ACR offers advanced functionality not shown here, including configuring retention policies, storage account details, and geo-replication options. For more details on these and more, see Pulumi's [Azure NextGen](https://www.pulumi.com/docs/reference/pkg/azure-nextgen/containerregistry/registry/) or [Azure](https://www.pulumi.com/docs/reference/pkg/azure/containerservice/registry/) API documentation, or [Azure's own product page](https://azure.microsoft.com/en-us/services/container-registry/).

Reviewer 2: ```suggestion
ACR offers advanced functionality not shown here, including configuring retention policies, storage account details, and geo-replication options. For more details on these and more, see Pulumi's [Azure NextGen](https://www.pulumi.com/docs/reference/pkg/azure-nextgen/containerregistry/registry/) or [Azure](https://www.pulumi.com/docs/reference/pkg/azure/containerservice/registry/) API documentation, or [Azure's product page](https://azure.microsoft.com/en-us/services/container-registry/).
```

> To view another cloud provider's registry details, [select a new cloud in the switcher above](#prepare-a-container-registry).

{{% /choosable %}}

{{% choosable cloud gcp %}}

<p></p>

### Google Container Registry (GCR)

Google Container Registry (GCR) enables you to store, manage, and secure your Docker container images. It includes built-in security vulnerability scanning and fast, high-availability access by storing images in regional private repositories across the world. This makes it easy to deploy custom application images to Google Kubernetes Engine (GKE), App Engine, or Cloud Run. Each account has a default registry per project per region, backed by Google Cloud Storage (GCS), and each registry is capable of storing many images and many versions.

#### Create a New Project

> To start, create a new project and [ensure it is configured to use your GCP account]({{< relref "/docs/intro/cloud-providers/gcp/setup" >}}), and then scaffold your project with the imports and overall program structure that we will fill in one piece at a time:

Reviewer 2: ```suggestion
To start, create a new project and [ensure it is configured to use your GCP account]({{< relref "/docs/intro/cloud-providers/gcp/setup" >}}), then scaffold your project with the imports and overall program structure that we will fill in one piece at a time:
```

```

{{< /choosable >}}

{{< choosable language go >}}

```go
package main

import (
    "encoding/base64"
    "errors"
    "strings"

    "github.com/pulumi/pulumi-docker/sdk/v2/go/docker"
    "github.com/pulumi/pulumi-gcp/sdk/v2/go/gcp/container"
    "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        // [Placeholder 1: Create a private GCR registry.]

        // [Placeholder 2: Get registry info (creds and endpoint).]

        // [Placeholder 3: Build and publish the container image.]
        return nil
    })
}
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Pulumi;
using Pulumi.Docker;
using Pulumi.Gcp.Container;

class Program
{
    static Task<int> Main() => Deployment.RunAsync(async () => {
        // [Placeholder 1: Create a private GCR registry.]

        // [Placeholder 2: Get registry info (creds and endpoint).]

        // [Placeholder 3: Build and publish the container image.]
    }
}
```

{{< /choosable >}}

#### Ensure Your GCR Registry is Ready

> Google Cloud automatically provisions a managed project-wide registry as we need it. To ensure our GCP registry is ready, we simply allocate a registry object, and then fetch the autogenerated repository URL that we can use for images:

Reviewer 2: ```suggestion
Google Cloud automatically provisions a managed project-wide registry as needed. To ensure our GCP registry is ready, we allocate a registry object, and then fetch the autogenerated repository URL that we can use for images:
```

const registryUrl = registry.id.apply(_ =>
    gcp.container.getRegistryRepository().then(reg => reg.repositoryUrl));
```

{{< /choosable >}}

{{< choosable language python >}}

```python
# Create a private GCR repository.
registry = gcp.container.Registry('my-registry')
registry_url = registry.id.apply(lambda _: gcp.container.get_registry_repository().repository_url)
```

{{< /choosable >}}

{{< choosable language go >}}

```go
// Create a private GCR registry.
registry, err := container.NewRegistry(ctx, "my-registry", nil)
if err != nil {
    return err
}
registryUrl := registry.ID().ApplyString(func(_ string) (string, error) {
    rep, err := container.GetRegistryRepository(ctx, nil)
    if err != nil {
        return "", err
    }
    return rep.RepositoryUrl, nil
})
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Create a private GCR registry.
var registry = new Registry("my-registry");
var registryUrl = registry.Id.Apply(async _ => {
    return (await GetRegistryRepository.InvokeAsync()).RepositoryUrl;
});
```

{{< /choosable >}}

Below we will append the container repository name to this registry URL.

#### Authenticate with the Google Cloud SDK

> GCP offers [several mechanisms](https://cloud.google.com/container-registry/docs/advanced-authentication) to authenticate with your registry, however, the most secure and preferred option is to use the Google Cloud CLI, `gcloud`, as a Docker credential helper. After [setting up gcloud on your client](https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud-helper), no explicit authentication is required in your code &mdash; instead, the ambient `gcloud` authentication settings will be used instead. As a result, we can leave the authentication information blank:

Reviewer 2: ```suggestion
GCP offers [several mechanisms](https://cloud.google.com/container-registry/docs/advanced-authentication) to authenticate to your registry; however, the most secure and preferred option is to use the Google Cloud CLI, `gcloud`, as a Docker credential helper. After [setting up gcloud on your client](https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud-helper), no explicit authentication is required in your code &mdash; instead, the ambient `gcloud` authentication settings will be used instead. As a result, we can leave the authentication information blank:
```

```javascript
// Get registry info (creds and endpoint).
var imageName = registryUrl.apply(url => `${url}/myapp`);
var registryInfo = undefined; // use gcloud for authentication.
```

{{< /choosable >}}

{{< choosable language typescript >}}

```typescript
// Get registry info (creds and endpoint).
const imageName = registryUrl.apply(url => `${url}/myapp`);
const registryInfo = undefined; // use gcloud for authentication.
```

{{< /choosable >}}

{{< choosable language python >}}

```python
# Get registry info (creds and endpoint).
image_name = registry_url.apply(lambda url: f'{url}/myapp')
registry_info = None # use gcloud for authentication.
```

{{< /choosable >}}

{{< choosable language go >}}

```go
// Get registry info (creds and endpoint).
imageName := pulumi.Sprintf("%s/myapp", registryUrl)
registryInfo := docker.ImageRegistryArgs{} // use gcloud for authentication.
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Get registry info (creds and endpoint).
var imageName = Output.Format($"{registryUrl}/myapp");
// Omit registryInfo and use gcloud for authentication.
```

{{< /choosable >}}

#### Learn More About GCR

> GCR uses Google Cloud Storage to store images, which may be configured separately, including configuring IAM and lifecycle policies. For more details on this, please refer to [Pulumi's Google Cloud API](https://www.pulumi.com/docs/reference/pkg/gcp/container/registry/) or [Google's own Container Registry](https://cloud.google.com/container-registry) documentation.

Reviewer 2: ```suggestion
GCR uses Google Cloud Storage to store images, which may be configured separately, including configuring IAM and lifecycle policies. For more details on this, please refer to [Pulumi's Google Cloud API](https://www.pulumi.com/docs/reference/pkg/gcp/container/registry/) or [Google's Container Registry](https://cloud.google.com/container-registry) documentation.
```

> To view another cloud provider's registry details, [select a new cloud in the switcher above](#prepare-a-container-registry).

{{% /choosable %}}

{{% choosable cloud digitalocean %}}

<p></p>

### DigitalOcean Container Registry

> DigitalOcean's Container Registry is an easy way to store and manage private container images for your applications for easy use with DigitalOcean's managed Kubernetes service.

Reviewer 2: ```suggestion
DigitalOcean's Container Registry is an easy way to store and manage private container images for your applications for use with DigitalOcean's managed Kubernetes service.
```

#### Create a New Project

> To start, create a new project and [ensure it is configured to use your DigitalOcean account]({{< relref "/docs/intro/cloud-providers/digitalocean/setup" >}}), and then scaffold your project with the imports and overall program structure that we will fill in one piece at a time:

Reviewer 2: ```suggestion
To start, create a new project and [ensure it is configured to use your DigitalOcean account]({{< relref "/docs/intro/cloud-providers/digitalocean/setup" >}}) and then scaffold your project with the imports and overall program structure that we will fill in one piece at a time:
```

var registry = new digitalocean.ContainerRegistry("my-reg", {
    subscriptionTierSlug: "starter",
});
```

{{< /choosable >}}

{{< choosable language typescript >}}

```typescript
// Create a private DigitalOcean Container Registry.
const registry = new digitalocean.ContainerRegistry("my-reg", {
    subscriptionTierSlug: "starter",
});
```

{{< /choosable >}}

{{< choosable language python >}}

```python
# Create a private DigitalOcean Container Registry.
repo = digitalocean.ContainerRegistry('my-reg',
    subscription_tier_slug='starter',
)
```

{{< /choosable >}}

{{< choosable language go >}}

```go
// Create a private DigitalOcean Container Registry.
registry, err := digitalocean.NewContainerRegistry(ctx, "my-reg",
    &digitalocean.ContainerRegistryArgs{
        // SubscriptionTierSlug: pulumi.String("starter"),
    })
if err != nil {
    return err
}
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Create a private DigitalOcean Container Registry.
var registry = new ContainerRegistry("my-reg", new ContainerRegistryArgs
{
    SubscriptionTierSlug = "starter"
});
```

{{< /choosable >}}

#### Authenticate with Temporary Credentials

> DigitalOcean supports generating temporary read or read/write credentials that Docker can use to authenticate with your new private container registry, which we'll now use. This resource returns a new credentials file similar to what the `docker login` command generates, and we can parse it and base64 decode its contents to discover the temporary username and password. Here we gather that plus the registry's URL endpoint in preparation for building and publishing the container image to the newly provisioned registry:

Reviewer 2: ```suggestion
DigitalOcean supports generating temporary read or read/write credentials that Docker can use to authenticate with your new private container registry, which we'll now use. This resource returns a new credentials file similar to what the `docker login` command generates, and we can parse it and base64 decode its contents to discover the temporary username and password. Here, we gather the username and password plus the registry's URL endpoint in preparation for building and publishing the container image to the newly provisioned registry:
```

}
```

{{< /choosable >}}

{{< choosable language csharp >}}

```csharp
// Fetch the Docker Hub auth info from config.
var config = new Pulumi.Config();
var username = config.Require("dockerUsername");
var password = config.RequireSecret("dockerPassword");

// Populate the registry info (creds and endpoint).
var imageName = $"{username}/myapp";
var registryInfo = new ImageRegistry
{
    Server = "docker.io",
    Username = username,
    Password = password,
};
```

{{< /choosable >}}

After writing this code, add your Docker username and password to your Pulumi project:

```bash
$ pulumi config set dockerUsername [YOUR USERNAME]
$ pulumi config set --secret dockerAccessToken [YOUR TOKEN]
```

There are of course other ways to configure this authentication information, however, the Pulumi secrets system ensures that your token is encrypted and safe to use.

> To view another cloud provider's registry details, [select a new cloud in the switcher above](#prepare-a-container-registry).

{{% /choosable %}}

At this stage, we can run `pulumi up` to check that the program works and to provision any necessary cloud resources.

## Build and Publish Your Container

> Now we are ready to build and publish your container image to the chosen registry. [The Docker provider's `Image` component]({{< relref "/docs/reference/pkg/docker" >}}) internally uses the Docker engine to perform the necessary steps to carry this out, including building, tagging, capturing and streaming container build logs as progress is made, and pushing the final result.

Reviewer 2: ```suggestion
Now we are ready to build and publish your container image to the chosen registry. [The Docker provider's `Image` component]({{< relref "/docs/reference/pkg/docker" >}}) internally uses the Docker engine to perform the necessary steps to carry this out, including building, tagging, capturing, and streaming container build logs as progress is made and pushing the final result.
```

        // Build and publish the container image.
        var image = new Image("my-image", new ImageArgs
        {
            Build = new DockerBuild { Context = "app" },
            ImageName = imageName,
            Registry = registryInfo,
        });

        // Export the base and specific version image name.
        return new Dictionary<string, object>
        {
            { "baseImageName", image.BaseImageName },
            { "fullImageName", image.ImageName },
        };
    });
}
```

{{< /choosable >}}

Now let's run `pulumi up`. If you haven't already done so, this will provision the cloud resources, as well as build/publish the container image. As the build runs, you will see Docker build output streamed to your terminal:

```bash
$ pulumi up
Updating (dev)

     Type                   Name            Status
 +   pulumi:pulumi:Stack    project-dev     created
 ... cloud-specific resources omitted ...
 +   └─ docker:image:Image  my-image        created

Outputs:
    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
    fullImageName: "...cloud-specific url.../my-repo-dc811b0:78d0fce7c2450c15a6153b6b11208fcb6b9edea7bb7ef3b7b6194f3fc101a170"
```

> Note also the base and versioned image URLs are exported as stack outputs. These are optional but have three benefits:

Reviewer 2: ```suggestion
Also note the base and versioned image URLs are exported as stack outputs. These are optional but have three benefits:
```

1. As we see here, the CLI will print them after each deployment.

> 2. The CLI can fetch them on-demand, e.g., 'pulumi stack output fullImageName', making it easy to script actress to these container images.

Reviewer 2: ```suggestion
2. The CLI can fetch them on-demand, e.g., 'pulumi stack output fullImageName', making it easy to script access to these container images.
```

> 3. The [Pulumi `StackReference` component]({{< relref "/docs/intro/concepts/organizing-stacks-projects#inter-stack-dependencies" >}}) can be used to depend on this stack from another, allowing up to build higher levels of infrastructure that consume these images.

Reviewer 2: ```suggestion
3. The [Pulumi `StackReference` component]({{< relref "/docs/intro/concepts/organizing-stacks-projects#inter-stack-dependencies" >}}) can be used to depend on this stack from another, allowing us to build higher levels of infrastructure that consume these images.
```

As an example of (2) in action, let's run the image locally using the exported image name:

```bash
$ pulumi stack output fullImageName
...cloud-specific url.../my-repo-dc811b0:78d0fce7c2450c15a6153b6b11208fcb6b9edea7bb7ef3b7b6194f3fc101a170
```

> Note that this requires that our client can pull from our provisioned registry, which may require cloud-specific authentication first.

> To redeploy changes to the `Dockerfile` or anything that image depends on, including application code, just rerun `pulumi up`. It will detect the differences, rebuild and repush the minimal layer changes necessary to update and retag the repository image, and then export the resulting image name.

Reviewer 2: ```suggestion
To redeploy changes to the `Dockerfile` or anything that image depends on, including application code, rerun `pulumi up`. It will detect the differences, rebuild and re-push the minimal layer changes necessary to update and retag the repository image, and then export the resulting image name.
```

```

And then rerun Pulumi -- notice how it detects the change, updates the image, and the versioned image hash changes (indicated by the tilde `~`):

```bash
$ pulumi up
Updating (dev)

     Type                   Name            Status
 +   pulumi:pulumi:Stack    project-dev     created
 ~   └─ docker:image:Image  my-image        updated

Outputs:
    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
  ~ fullImageName: "...cloud-specific url.../my-repo-dc811b0:78d0fce7c2450c15a6153b6b11208fcb6b9edea7bb7ef3b7b6194f3fc101a170"
```

The Pulumi Docker `Image` component supports a number of additional options to control its behavior, including passing build arguments, environment variables, extra options for the Docker build, as well as various controls for image caching. For full details on each of these, refer to [the API documentation]({{< relref "/docs/reference/pkg/docker/image" >}}).

## Consume the Container Image

The same container image URLs exported above can be used as inputs to other resources, including infrastructure that will run your container inside of a container orchestration system such as Kubernetes, Amazon ECS, and so on.

> > This article assumes you already have a containerized environment to deploy to, like a Kubernetes cluster, and have [configured your project accordingly]({{< relref "/docs/intro/cloud-providers/kubernetes/setup" >}}). If not, you can provision one using Pulumi first. Pulumi supports many clouds and infrastructure resources, but here are a few starting points to get up and running with: [AWS Elastic Container Service ECS]({{< relref "/docs/guides/crosswalk/aws/ecs" >}}), [AWS Elastic Kubernetes Service (EKS)]({{< relref "/docs/guides/crosswalk/aws/eks" >}}), [Azure Kubernetes Service (AKS)]({{< relref "/docs/tutorials/kubernetes/aks" >}}), [Google Cloud Kubernetes Engine (GKE)]({{< relref "/docs/tutorials/kubernetes/gke" >}}), [DigitalOcean Kubernetes](https://www.digitalocean.com/community/tutorials/how-to-manage-digitalocean-and-kubernetes-infrastructure-with-pulumi).

Reviewer 2: ```suggestion
> This article assumes you already have a containerized environment to deploy to, like a Kubernetes cluster, and have [configured your project accordingly]({{< relref "/docs/intro/cloud-providers/kubernetes/setup" >}}). If not, you can provision one using Pulumi first. Pulumi supports many clouds and infrastructure resources, but here are a few starting points to get up and running with: 

- [AWS Elastic Container Service ECS]({{< relref "/docs/guides/crosswalk/aws/ecs" >}}), 
- [AWS Elastic Kubernetes Service (EKS)]({{< relref "/docs/guides/crosswalk/aws/eks" >}}), 
- [Azure Kubernetes Service (AKS)]({{< relref "/docs/tutorials/kubernetes/aks" >}}), 
- [Google Cloud Kubernetes Engine (GKE)]({{< relref "/docs/tutorials/kubernetes/gke" >}}), 
- [DigitalOcean Kubernetes](https://www.digitalocean.com/community/tutorials/how-to-manage-digitalocean-and-kubernetes-infrastructure-with-pulumi).
```

> In this example, we will demonstrate deploying our Nginx web server as a load balanced service within Kubernetes. To do so, we'll declare our Kubernetes configuration, right inside of our existing program defined above, and export its resulting IP address:

Reviewer 2: ```suggestion
This example demonstrates deploying our Nginx web server as a load balanced service within Kubernetes. To do so, we'll declare our Kubernetes configuration, right inside of our existing program defined above, and export its resulting IP address:
```

{{< chooser language "javascript,typescript,python,go,csharp" / >}}

{{< choosable language javascript >}}

```javascript
// [Previous imports...]
var k8s = require("@pulumi/kubernetes");

> // [Registry, build, etc, code from above...]

Reviewer 2: ```suggestion
// [Registry, build, etc., code from above...]
```

        };
    }
}
```

{{< /choosable >}}

Notice here that we are referring to the resulting image name from our service specification. All we need to do run a 'pulumi up' and Pulumi will now build, publish, and consume our application's container image from within our Kubernetes cluster:

```bash
$ pulumi up
Updating (dev)

     Type                              Name            Status
 +   pulumi:pulumi:Stack               project-dev     created
 ... prior output omitted ...
 +   ├─ kubernetes:apps/v1:Deployment  app-dep         created
 +   └─ kubernetes:core/v1:Service     app-svc         created

Outputs:
    appIp        : "34.82.32.166"
    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
    fullImageName: "...cloud-specific url.../my-repo-dc811b0:fcbbd958636e4f8ac02568db7dbf8b89a043e210e917432cf12649aaedfa266c"
```

If this is the first time you've run `pulumi up`, all of the prior resources will be provisioned first, and in any case, the Kubernetes resources will follow.

Now that our deployment has completed, we can curl the endpoint:

```bash
$ curl $(pulumi stack output appIp)
<h1>Hello, World -- from Pulumi!</h1>
```

> This image name is the versioned one, meaning that any time a new version is created, Pulumi will detect that it needs to update the Kubernetes deployment which references it, triggering a rollout of the new deployed image. As such, we just need to run `pulumi up` anytime the application or its `Dockerfile` changes and Pulumi will detect and trigger a deployment to Kubernetes.

Reviewer 2: ```suggestion
This image name is the versioned one, meaning that any time a new version is created, Pulumi will detect that it needs to update the Kubernetes deployment, which references it, triggering a rollout of the new deployed image. As such, we need to run `pulumi up` anytime the application or its `Dockerfile` changes, and Pulumi will detect and trigger a deployment to Kubernetes.
```

    /usr/share/nginx/html/index.html
```

And then rerun Pulumi -- notice that both the image _and_ the Kubernetes deployment has changed this time:

```bash
$ pulumi up
Updating (dev)

     Type                              Name            Status      Info
 +   pulumi:pulumi:Stack               project-dev     created
 ~   ├─ docker:image:Image             my-image        updated
 ~   └─ kubernetes:core/v1:Deployment  app-dep         updated     [diff: ~spec]

Outputs:
    appIp        : "34.82.32.166"
    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
    fullImageName: "...cloud-specific url.../my-repo-dc811b0:36919b56f6b2dedfa879815443c9964d17a89506caed25025328b1d0204a083f"
```

Finally, if we re-curl the endpoint, we will now see that the updated container is now live!

```
$ curl $(pulumi stack output appIp)
<h1>Hello, World -- from Pulumi and Kubernetes!</h1>
```

And there we go: we have gone from a `Dockerfile` to a published container image in a container registry of our choosing, which has been used to spin up a load balanced Kubernetes service, and which is now ready for fully automated continuous deployments.

## Wrapping Up

> In this article, we've seen how easy it is to build, publish, and use container images in many popular public and private container registry options. We have seen how to provision new registries using infrastructure as code and how easy it is to trigger deployments of application updates by simply running a single `pulumi up` command.

Reviewer 2: ```suggestion
In this article, we've seen how easy it is to build, publish, and use container images in many popular public and private container registry options. We have seen how to provision new registries using infrastructure as code and how easy it is to trigger deployments of application updates by running a single `pulumi up` command.
```

> All of these steps were manually run from a CLI, however, a natural next step is to [wire the entire process up to a CI/CD system]({{< relref "/docs/guides/continuous-delivery" >}}) such as [GitHub Actions]({{< relref "/docs/guides/continuous-delivery/github-actions" >}}), [GitLab Pipelines]({{< relref "/docs/guides/continuous-delivery/gitlab-ci" >}}), [Jenkins]({{< relref "/docs/guides/continuous-delivery/jenkins" >}}), [Spinnaker]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}), or [one of the many available options]({{< relref "/docs/guides/continuous-delivery" >}}, so that you can deploy continuously as you merge code. For even more advanced scenarios, we might want to build a custom program such as a CLI that uses the [Automation API]({{< relref "/blog/automation-api" >}}) to perform these actions behind a simpler, purpose-built interface.

Reviewer 2: ```suggestion
All of these steps were manually run from a CLI; however, a natural next step is to [wire the entire process up to a CI/CD system]({{< relref "/docs/guides/continuous-delivery" >}}) such as [GitHub Actions]({{< relref "/docs/guides/continuous-delivery/github-actions" >}}), [GitLab Pipelines]({{< relref "/docs/guides/continuous-delivery/gitlab-ci" >}}), [Jenkins]({{< relref "/docs/guides/continuous-delivery/jenkins" >}}), [Spinnaker]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}), or [one of the many available options]({{< relref "/docs/guides/continuous-delivery" >}}, so that you can deploy continuously as you merge code. For even more advanced scenarios, we might want to build a custom program such as a CLI that uses the [Automation API]({{< relref "/blog/automation-api" >}}) to perform these actions behind a simpler, purpose-built interface.
```

> Although we've shown a very simple set of infrastructure resources, it would be natural to extend these examples by provisioning other ancillary services that your application needs, including databases, pub/sub topics, queues, metrics and dashboards, and more. Using an infrastructure as code approach to building, publishing, and consuming your container images means you can incrementally add on such infrastructure to the base code shown above, and reference them from your containers easily with automatic dependency tracking.

Reviewer 2: ```suggestion
Although we've shown a very simple set of infrastructure resources, it would be natural to extend these examples by provisioning other ancillary services that your application needs, including databases, pub/sub topics, queues, metrics and dashboards, and more. Using an infrastructure as code approach to building, publishing, and consuming your container images means you can incrementally add on such infrastructure to the base code shown above and reference them from your containers easily with automatic dependency tracking.
```
